[flang][debug] Add support for common blocks.

abidh opened the pull request:
This PR adds debug support for common block in flang. As variable which are part of a common block don't have a special marker to recognize them, we use the following check to find them.

%0 = fir.address_of(@a)
%1 = fir.convert %0
%2 = fir.coordinate_of %1, %c0
%3 = fir.convert %2
%4 = fircg.ext_declare %3

If the memref of a fircg.ext_declare points to a fir.coordinate_of and that in turn points to an fir.address_of (ignoring immediate fir.convert) then we assume that it is a common block variable. The fir.address_of gives us the global symbol which is the storage for common block and fir.coordinate_of provides the offset in this storage.

The debug hierarchy looks like as

subroutine f3
  integer :: x, y
  common /a/ x, y
end subroutine

@a_ = global { ... } { ... }, !dbg !26, !dbg !28

!23 = !DISubprogram(name: "f3"...)
!24 = !DICommonBlock(scope: !23, name: "a", ...)
!25 = !DIGlobalVariable(name: "x", scope: !24 ...)
!26 = !DIGlobalVariableExpression(var: !25, expr: !DIExpression())
!27 = !DIGlobalVariable(name: "y", scope: !24 ...)
!28 = !DIGlobalVariableExpression(var: !27, expr: !DIExpression(DW_OP_plus_uconst, 4))

This required following changes:

1. Instead of using DIGlobalVariableAttr in the FusedLoc of GlobalOp, we use DIGlobalVariableExpressionAttr. This allows us the generate the DIExpression where we have the information.

2. Previously, only one DIGlobalVariableExpressionAttr could be linked to one global op. I recently removed this restriction in mlir. To make use of it, we add an ArrayAttr to the FusedLoc of a GlobalOp. This allows us to pass multiple DIGlobalVariableExpressionAttr.

3. I was depending on the name of global for the name of the common block. The name gets a '_' appended. I could not find a utility function in flang to remove it so I have to brute force it.

abidh committed:
[flang][debug] Add support for common blocks.

This PR adds debug support for common block in flang. As variable which
are part of a common block don't have a special marker to recognize them,
we use the following check to find them.

%0 = fir.address_of(@a)
%1 = fir.convert %0
%2 = fir.coordinate_of %1, %c0
%3 = fir.convert %2
%4 = fircg.ext_declare %3

If the memref of a fircg.ext_declare points to a fir.coordinate_of and
that in turn points to an fir.address_of (ignoring immediate fir.convert)
then we assume that it is a common block variable. The fir.address_of
gives us the global symbol which is the storage for common block and
fir.coordinate_of provides the offset in this storage.

The debug hierarchy looks like as

subroutine f3
  integer :: x, y
  common /a/ x, y
end subroutine

@a_ = global { ... } { ... }, !dbg !26, !dbg !28

!23 = !DISubprogram(name: "f3"...)
!24 = !DICommonBlock(scope: !23, name: "a", ...)
!25 = !DIGlobalVariable(name: "x", scope: !24 ...)
!26 = !DIGlobalVariableExpression(var: !25, expr: !DIExpression())
!27 = !DIGlobalVariable(name: "y", scope: !24 ...)
!28 = !DIGlobalVariableExpression(var: !27, expr: !DIExpression(DW_OP_plus_uconst, 4))

This required following changes:

1. Instead of using DIGlobalVariableAttr in the FusedLoc of GlobalOp, we
use DIGlobalVariableExpressionAttr. This allows us the generate the
DIExpression where we have the information.

2. Previously, only one DIGlobalVariableExpressionAttr could be linked to
one global op. I recently removed this restriction in mlir. To make use
of it, we add an ArrayAttr to the FusedLoc of a GlobalOp. This
allows us to pass multiple DIGlobalVariableExpressionAttr.

3. I was depending on the name of global for the name of the common block.
The name gets a '_' appended. I could not find a utility function
in flang to remove it so I have to brute force it.

tblah commented:
I think you could make the detection more robust by ensuring that the linkage attribute of the `fir.global` is `"common"`.

tblah commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -90,6 +105,67 @@ bool debugInfoIsAlreadySet(mlir::Location loc) {
+              symbolTable->lookup<fir::GlobalOp>(sym.getRootReference())) {
+
+        unsigned line = getLineFromLoc(global.getLoc());
+        std::string commonName(sym.getRootReference().str());
```
ultra-nit: Creating a `std::string` leads to unnecessary heap allocation. A `SymbolRefAttr` can also be converted into an `llvm::StringRef`, which can then be used as an argument type to `getOrCreateCommonBlockAttr`.

`StringRef::drop_back` just creates a `StringRef` with a shorter length. It does not modify the source buffer.

abidh replied:
Thanks for the suggestion. I have used llvm::StringRef now.

tblah commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -90,6 +105,67 @@ bool debugInfoIsAlreadySet(mlir::Location loc) {
+  mlir::MLIRContext *context = &getContext();
+  mlir::OpBuilder builder(context);
+  std::optional<std::int64_t> optint;
+  auto op = declOp.getMemref().getDefiningOp();
```
nit: Usual flang style is to spell out types where these are not stated in the initializer for the variable.

For example, `auto sym = addrOfOp.getSymbol()` should spell out the type of `sym`, but `auto conOp = mlir::dyn_cast_if_present<fir::convertOp>(op)` already makes the type clear.

abidh replied:
Thanks for the catch. Fixed now.

abidh commented:
> I think you could make the detection more robust by ensuring that the linkage attribute of the `fir.global` is `"common"`.

I actually had this check but later noticed during my testing that sometime that global would not have "common" linkage. IIRC it was when common block is declared inside a module.

tblah commented:
> > I think you could make the detection more robust by ensuring that the linkage attribute of the `fir.global` is `"common"`.
> 
> I actually had this check but later noticed during my testing that sometime that global would not have "common" linkage. IIRC it was when common block is declared inside a module.

Ahh good catch. My concern is that operation sequence you are matching looks quite generic. I wonder if we should add a unit attribute to fir.global which indicates whether it comes from a common block and set that during lowering?

abidh committed:
Handle review comments.

abidh committed:
Remove an unnecessary str() call.

jeanPerier commented:
> I actually had this check but later noticed during my testing that sometime that global would not have "common" linkage. IIRC it was when common block is declared inside a module.

Can you give an example, I think a common block global without common linkage is most likely a bug (which is possible).

However, I would not push on assuming that common linkage imply common block, we may want to use that in other cases. What about adding an attribute on the global instead with the common block name so that you do not have to care about the assembly name, like `fir.common_block = "a"`?

I also feel like adding an attribute to the declare with the offset and global name may be a lot more robust than pattern matching the computation (something like `fir.declare ...... {fir.in_common_block = "a", fir.common_block_offset = 12}`) would be cleaner. It duplicate information a bit, but I think retrieving information in DAGs outside of optimization purposes is quickly brittle.

If those option sounds good to you, I am happy to extend FIR here.

abidh commented:
> > I actually had this check but later noticed during my testing that sometime that global would not have "common" linkage. IIRC it was when common block is declared inside a module.
> 
> Can you give an example, I think a common block global without common linkage is most likely a bug (which is possible).
> 
> However, I would not push on assuming that common linkage imply common block, we may want to use that in other cases. What about adding an attribute on the global instead with the common block name so that you do not have to care about the assembly name, like `fir.common_block = "a"`?
> 

Here is the example code and the generated fir.

```
module helper
  implicit none
  integer :: x = 42, xa = 52
  common /a/ x, xa
end module helper


  fir.global @a_ {alignment = 4 : i64} : tuple<i32, i32> {
    %0 = fir.zero_bits tuple<i32, i32>
    %c42_i32 = arith.constant 42 : i32
    %1 = fir.insert_value %0, %c42_i32, [0 : index] : (tuple<i32, i32>, i32) -> tuple<i32, i32>
    %c52_i32 = arith.constant 52 : i32
    %2 = fir.insert_value %1, %c52_i32, [1 : index] : (tuple<i32, i32>, i32) -> tuple<i32, i32>
    fir.has_value %2 : tuple<i32, i32>
  }

```
> I also feel like adding an attribute to the declare with the offset and global name may be a lot more robust than pattern matching the computation (something like `fir.declare ...... {fir.in_common_block = "a", fir.common_block_offset = 12}`) would be cleaner. It duplicate information a bit, but I think retrieving information in DAGs outside of optimization purposes is quickly brittle.
> 
> If those option sounds good to you, I am happy to extend FIR here.

I will agree to that. I had this concern that we are depending on pattern matching which is quite fragile. Having this information in IR would be great. Another item on my wish list is having the position of the argument to function/subroutine on the declare op which is will save us from doing similar pattern matching for CHARACTER type arguments.

jinisusan reviewed (commented):
A couple of comments: 

* It might be good to have the common block variables tagged with the  "DW_AT_external" attribute in line with what is done by other fortran compilers. 
* Also, it might be good to include the linkage name attribute while generating DICommonBlock

jieljiel commented:
Hi @abidh, I'm visiting this PR and wondering if you would like to share your plan about how to proceed with this support for common blocks? Is the FIR extension that @jeanPerier suggested a prerequiste?

jieljiel commented:
> > Hi @abidh, I'm visiting this PR and wondering if you would like to share your plan about how to proceed with this support for common blocks? Is the FIR extension that @jeanPerier suggested a prerequiste?
> 
> Hi @jieljiel
> 
> The FIR extension is the best way to detect the common block but the current approach also works. I was wondering that we can use this PR for now as it fills a missing hole in the debug support. When the extension in FIR to detect common block is available in future, we can use that and remove the DAG walking. If @jeanPerier thinks this is OK then I can rebase/update the PR and address any other comments.

@jeanPerier Would you mind letting @abidh 's PR in and then he will make adjustment based on FIR extensions later.

@abidh Would you be able to implement the required FIR extension, or you need @jeanPerier 's help working on that?

jeanPerier commented:
@jeanPerier Would you mind letting @abidh 's PR in and then he will make adjustment based on FIR extensions later.

Yes, you can go ahead and update/merge this and I will ping you when I have the FIR changes.

jeanPerier commented:
>> Can you give an example, I think a common block global without common linkage is most likely a bug (which is possible).
> Here is the example code and the generated fir.

I never replied to that, thanks for the example. It is actually correct and expected to not give the "common" attribute to global symbols for common block with an initial value because it is _the_ definition of the symbol (in other compilation units, it cannot be given initial values and will have the common linkage).

LLVM IR actually forbids common linkage on a global with a non zero initial value (from [LLVM ref manual](https://llvm.org/docs/LangRef.html#linkage-types): _common symbols may not have an explicit section, must have a zero initializer, and may not be marked constant_.)

abidh committed:
Handle review comments (2).

Set isLocalToUnit to false to get DW_AT_external=true in DWARF.

abidh commented:
> > @jeanPerier Would you mind letting @abidh 's PR in and then he will make adjustment based on FIR extensions later.
> 
> Yes, you can go ahead and update/merge this and I will ping you when I have the FIR changes.

Hi @jeanPerier 
Your comment suggested that this PR is ok to go in. Can you kindly approve if it is still the case. Thanks.

jeanPerier commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -2994,11 +2994,10 @@ struct GlobalOpConversion : public fir::FIROpConversion<fir::GlobalOp> {
-            global.getContext(), gvAttr, mlir::LLVM::DIExpressionAttr()));
+      if (auto gvExprAttr = mlir::dyn_cast_if_present<mlir::ArrayAttr>(
+              fusedLoc.getMetadata())) {
+        for (auto attr : gvExprAttr.getAsRange<mlir::Attribute>())
```
Should there be  a `if(auto dbgAttr = mlir::dyn_cast<DIGlobalVariableExpressionAttr>(attr))` to ensure these is indeed debug info?

jeanPerier replied:
Thanks for the change and link.

jeanPerier reviewed (approved):
Small question, LGTM otherwise, thanks.

abidh committed:
Handle review comments (3).

Add a check to ensure we do have `DIGlobalVariableExpressionAttr` in the `ArrayAttr`.
